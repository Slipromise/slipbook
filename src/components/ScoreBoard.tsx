import React, { useMemo } from "react";
import Container from "react-bootstrap/esm/Container";
import Stack from "react-bootstrap/esm/Stack";
import styles from "@/styles/components/ScoreBoard.module.scss";
import dayjs from "dayjs";
import { useSpring, animated, AnimationConfig } from "@react-spring/web";

type Props = {
  aTitle: string;
  bTitle: string;
  aScore: number;
  bScore: number;
  countDownTime?: number;
  countDownFormat?: "HH:mm" | "HH:mm:ss" | "mm:ss" | "ss.SSS";
  subTitle?: string;
};

function ScoreBoard({
  aTitle,
  aScore,
  bTitle,
  bScore,
  countDownTime,
  countDownFormat = "mm:ss",
  subTitle,
}: Props) {
  const containerStyle = useMemo<React.CSSProperties>(
    () =>
      ({
        "--percent": `${(aScore / (aScore + bScore)) * 100}%`,
      }) as React.CSSProperties,
    [aScore, bScore]
  );

  const scores = useSpring<{ aScore: number; bScore: number }>({
    aScore,
    bScore,
    config: {
      duration: 300,
    } as AnimationConfig,
  });

  return (
    <Container className={styles.container} style={containerStyle}>
      <Stack direction="horizontal" data-team style={{ flexGrow: aScore }}>
        <span data-team-name data-testId="homeTitle">
          {aTitle}
        </span>
        <animated.span data-team-score data-testId="homeScore">
          {scores.aScore.to((n) => Math.floor(n))}
        </animated.span>
      </Stack>
      <Stack direction="horizontal" data-team style={{ flexGrow: bScore }}>
        <animated.span data-team-score data-testId="awayScore">
          {scores.bScore.to((n) => Math.floor(n))}
        </animated.span>
        <span data-team-name data-testId="awayTitle">
          {bTitle}
        </span>
      </Stack>
      {countDownTime && (
        <span data-countdown>
          {dayjs.duration(countDownTime).format(countDownFormat)}
        </span>
      )}
      {subTitle && <span data-subtitle>{subTitle}</span>}
    </Container>
  );
}

export default ScoreBoard;
