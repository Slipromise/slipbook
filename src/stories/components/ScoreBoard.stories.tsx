import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import ScoreBoard from "../../components/ScoreBoard";

const meta: Meta<typeof ScoreBoard> = {
  component: ScoreBoard,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof ScoreBoard>;

export const Normal: Story = {
  args: {
    aTitle: "home",
    bTitle: "away",
    aScore: 7,
    bScore: 10,
  },
};

export const WithCountdown: Story = {
  args: {
    aTitle: "home",
    bTitle: "away",
    aScore: 10,
    bScore: 7,
    countDownTime: 30000,
  },
};

export const WithSubtitle: Story = {
  args: {
    aTitle: "home",
    bTitle: "away",
    aScore: 10,
    bScore: 7,
    subTitle: "1 - 3",
  },
};
