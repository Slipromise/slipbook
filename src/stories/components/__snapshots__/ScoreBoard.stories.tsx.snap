// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/ScoreBoard Normal smoke-test 1`] = `
<div class="ScoreBoard_container__Kf5rs container"
     style="--percent: 41.17647058823529%;"
>
  <div data-team="true"
       class="hstack"
       style="flex-grow: 7;"
  >
    <span data-team-name="true"
          data-testid="homeTitle"
    >
      home
    </span>
    <span data-team-score="true"
          data-testid="homeScore"
    >
      7
    </span>
  </div>
  <div data-team="true"
       class="hstack"
       style="flex-grow: 10;"
  >
    <span data-team-score="true"
          data-testid="awayScore"
    >
      10
    </span>
    <span data-team-name="true"
          data-testid="awayTitle"
    >
      away
    </span>
  </div>
</div>
`;

exports[`stories/components/ScoreBoard WithCountdown smoke-test 1`] = `
<div class="ScoreBoard_container__Kf5rs container"
     style="--percent: 58.82352941176471%;"
>
  <div data-team="true"
       class="hstack"
       style="flex-grow: 10;"
  >
    <span data-team-name="true"
          data-testid="homeTitle"
    >
      home
    </span>
    <span data-team-score="true"
          data-testid="homeScore"
    >
      10
    </span>
  </div>
  <div data-team="true"
       class="hstack"
       style="flex-grow: 7;"
  >
    <span data-team-score="true"
          data-testid="awayScore"
    >
      7
    </span>
    <span data-team-name="true"
          data-testid="awayTitle"
    >
      away
    </span>
  </div>
  <span data-countdown="true">
    00:30
  </span>
</div>
`;

exports[`stories/components/ScoreBoard WithSubtitle smoke-test 1`] = `
<div class="ScoreBoard_container__Kf5rs container"
     style="--percent: 58.82352941176471%;"
>
  <div data-team="true"
       class="hstack"
       style="flex-grow: 10;"
  >
    <span data-team-name="true"
          data-testid="homeTitle"
    >
      home
    </span>
    <span data-team-score="true"
          data-testid="homeScore"
    >
      10
    </span>
  </div>
  <div data-team="true"
       class="hstack"
       style="flex-grow: 7;"
  >
    <span data-team-score="true"
          data-testid="awayScore"
    >
      7
    </span>
    <span data-team-name="true"
          data-testid="awayTitle"
    >
      away
    </span>
  </div>
  <span data-subtitle="true">
    1 - 3
  </span>
</div>
`;
